File: Agents/Azure-Agent-Skills-In-Action/skill-demos/sdk-python/06_azure_servicebus.py
"""
TRIPLE:
  Skill: azure-servicebus-py
  Prompt: "Using azure-servicebus-py skill, write a Python script that sends a document-ingestion
           message to a Service Bus queue using DefaultAzureCredential, then receives it with
           peek-lock and settles (complete or dead-letter on parse failure)."
  Deliverable: This file — runnable Python script

Source: https://github.com/microsoft/skills/tree/main/.github/plugins/azure-sdk-python/skills/azure-servicebus-py
"""
import json
import os
from azure.identity import DefaultAzureCredential
from azure.servicebus import ServiceBusClient, ServiceBusMessage

namespace = os.environ["SERVICEBUS_NAMESPACE"]  # <namespace>.servicebus.windows.net
queue_name = os.environ.get("SERVICEBUS_QUEUE", "document-ingestion")
credential = DefaultAzureCredential()

client = ServiceBusClient(fully_qualified_namespace=namespace, credential=credential)

# Send
with client.get_queue_sender(queue_name) as sender:
    msg = ServiceBusMessage(
        body=json.dumps({"blob_url": "https://storage/container/doc.pdf", "action": "index"}),
        content_type="application/json",
        subject="document-ingestion",
    )
    sender.send_messages(msg)
    print(f"Sent message to queue '{queue_name}'")

# Receive with peek-lock
with client.get_queue_receiver(queue_name, max_wait_time=10) as receiver:
    for msg in receiver:
        try:
            body = json.loads(str(msg))
            print(f"Received: {body}")
            receiver.complete_message(msg)
            print("  → completed")
        except json.JSONDecodeError:
            receiver.dead_letter_message(msg, reason="parse_failure", error_description="Invalid JSON")
            print("  → dead-lettered (invalid JSON)")

File: Agents/Azure-Agent-Skills-In-Action/skill-demos/sdk-typescript/06_azure_servicebus.ts
// TRIPLE:
//   Skill: azure-servicebus-ts
//   Prompt: "Using azure-servicebus-ts skill, write TypeScript code that sends a message, receives with peek-lock, and completes or dead-letters."
//   Deliverable: This file
//   Source: https://github.com/microsoft/skills/tree/main/.github/plugins/azure-sdk-typescript/skills/azure-servicebus-ts

import { ServiceBusClient, ServiceBusReceivedMessage } from "@azure/service-bus";
import { DefaultAzureCredential } from "@azure/identity";

async function main(): Promise<void> {
  const fullyQualifiedNamespace = process.env["AZURE_SERVICEBUS_NAMESPACE"]!;
  const queueName = process.env["AZURE_SERVICEBUS_QUEUE"] ?? "skill-demo-queue";
  const credential = new DefaultAzureCredential();

  const sbClient = new ServiceBusClient(fullyQualifiedNamespace, credential);

  // Send a message
  const sender = sbClient.createSender(queueName);
  await sender.sendMessages({
    body: { action: "process", timestamp: new Date().toISOString() },
    contentType: "application/json",
    subject: "skill-demo",
  });
  console.log("Message sent.");
  await sender.close();

  // Receive with peek-lock (default mode)
  const receiver = sbClient.createReceiver(queueName);
  const messages: ServiceBusReceivedMessage[] = await receiver.receiveMessages(1, {
    maxWaitTimeInMs: 5000,
  });

  for (const msg of messages) {
    console.log(`Received: ${JSON.stringify(msg.body)}`);
    try {
      // Process message then complete
      await receiver.completeMessage(msg);
      console.log("Message completed.");
    } catch {
      // On failure, dead-letter the message
      await receiver.deadLetterMessage(msg, {
        deadLetterReason: "ProcessingFailed",
        deadLetterErrorDescription: "Demo: simulated failure",
      });
      console.log("Message dead-lettered.");
    }
  }

  await receiver.close();
  await sbClient.close();
}

main().catch(console.error);
